ci/run_example_tests.sh

```diff
@@ -0,0 +1,470 @@
+#!/bin/bash -E
+
+FAILED=()
+
+
+trap_errors () {
+    local frame=0 LINE SUB FILE
+    set +v
+    while read -r LINE SUB FILE < <(caller "$frame"); do
+	if [ "$frame" -ne "0" ]; then
+	    FAILED+=("  > ${SUB}@ ${FILE} :${LINE}")
+	else
+	    FAILED+=("${SUB}@ ${FILE} :${LINE}")
+	fi
+	((frame++))
+    done
+    set -v
+}
+
+trap trap_errors ERR
+
+run_log () {
+    local name
+    name="$1"
+    shift
+    echo -e "\n> [${name}] ${*}"
+}
+
+run_log "$(whoami)" "Running tests"
+run_log "$(whoami)" "Set user umask  ($ENVOYDEV_UMASK)"
+umask "$ENVOYDEV_UMASK"
+
+show_user_env () {
+    run_log "$(whoami)" "User env"
+    id
+    echo "umask = $(umask)"
+    echo "pwd = $(pwd)"
+}
+
+clone_envoy_repo () {
+    local git_args
+    if [ -d envoy ]; then
+	run_log "$(whoami)" "Testing mounted envoy directory"
+    else
+	read -ra git_args <<< "$ENVOY_REPO"
+	run_log "$(whoami)" "Clone envoy repository ($ENVOY_REPO)"
+	git clone "${git_args[@]}"
+    fi
+}
```

Comment:
You shouldn't do this, use CI checked-out one.

Reply:
it only does this if the already checked out repo isnt mounted - this allows it to be used against other branches

i can remove, but its kinda useful

Reply:
If you want to check against other branch, just checkout it? Are you saying you want to check branches other than where envoy is built from? Then the argument should be the docker image tag.

Reply:
i guess mostly because ive been running this in an environment that does not already have a checkout of the envoy repo (ie on travis) it has been much easier to do this:

```
export ENVOY_REPO=...

docker run --rm phlax/envoy-examples 

```

rather than having to checkout the repo and mount it into the container

im happy to remove

Reply:
Let's start small, and  use what we need for CI first. Non-used test code / script usually get out of date and stop working.

Reply:
see above comment about umask - as that is relevant here  too

Reply:
my original plan here was to do:

```
export ENVOY_REPO="-b $(git branch --show-current) $(git remote get-url origin)"
```
i abandoned that for 2 reasons - firstly because the repo is already checked out in `ci`, but also because the git `remote` is (i think at least on azure) an ssh remote  - which wont work without a bit more effort

the problem i now realise is that the requirement to checkout the repo again (and use `d-in-d`) is to ensure that the checked out repo has exactly the permmissions that a user with a strict `umask` would have

Reply:
Let's not worry about umask for now? Aren't we already copying configs inside docker?

Reply:
> Let's not worry about umask for now? 

really?

i think it makes sense to test for this properly if the expectation is for this to work

> Aren't we already copying configs inside docker?

yep, but i think the point about adding (these) tests is to ensure that the docs continue to work

you can have whatever bits of the code you like, but it seems to me like we are adding a lot of code to then not test the the very thing the tests were used to fix.

i also think that the ci scripts should run locally - i would not want to run this locally outside of a container.

Reply:
@phlax My point here is to keep this PR minimum, let CI test what examples we have today, without adding many stuff, these kcan be in follow up PRs.

re: DinD, I don't personally run scripts in a container and neither most of our CI do. We can add support for that but should be optional.

Reply:
> @phlax My point here is to keep this PR minimum, let CI test what examples we have today, without adding many stuff, these kcan be in follow up PRs.

ill remove the clone code etc - and make the script work entirely standalone

at risk of sounding repetitive, and having just used these tests extensively to fix this particular problem - im surprised that we dont want to ensure it stays fixed

> re: DinD, I don't personally run scripts in a container and neither most of our CI do. We can add support for that but should be optional.

some things run in `ci/run_envoy_docker.sh` - and in my testing with building docs this  was necessary for the build script to work. Perhaps these are build tasks rather than ci, but im  not sure if that matters

the point about using `d-in-d` tho is 2fold
- we want as clean an env as possible so we are actually testing something like an end user's setup
- running all of these tests builds/creates (not properly namespaced) ~20-30 images. This takes up quite a bit of space, and is pita to sit and pick back out

Reply:
> at risk of sounding repetitive, and having just used these tests extensively to fix this particular problem - im surprised that we dont want to ensure it stays fixed

Thanks, yes I understand why it exists and I'm not saying we don't want, but not in this PR. Having a script other than CI checkout doing it having a risk that we're testing examples in master (or other branch that the binary is built being checked out), because the CI checkout is fairly complicated and subject to race (because it always use `pull/<PR>/merge` which is a generated commit).

Reply:
If DinD is necessary I'm OK leaving it, but I do need a way to run without it, so I can debug the examples with CI

Reply:
cool - ive removed it for now and shifted the code for it back to my repo - so i can continue testing.

yep, lets consider wrapping this test in a separate pr